File: KE2/latter term/kadai3.py
# coding: utf-8
import numpy as np
import math
# from numba import jit

DIM = 196
CTGRY = 1
DATASIZE = 180

# return data[DIM][DIM]
def fileRead_sigma(fname):
    f = open(fname,'r')
    # 配列の静的確保 data[DIM][DIM]
    data = [[0 for i in range(DIM)] for j in range(DIM)]
    for i in range(DIM):
        line = f.readline().split('\t')
        for j in range(DIM):
            data[i][j] = float(line[j])
    f.close()
    return data

#data[DIM][DIM]
def fileWrite(fname,data):
    with open(fname, mode='w') as f:
        for i in data:
            for j in i:
                f.write(str(j)+"\t")
            f.write('\n')

# 行列のクラス
class Matrix:
    # data[[]]
    def __init__(self,data):
        self.matrix = data             # ndarry[[]]
        self.row    = data.shape[0]    # row
        self.column = data.shape[1]    # column

# ヤコビ法計算クラス  < Matrix
class Jacobi(Matrix):
    def __init__(self,matrix):
        super().__init__(matrix)
        self.k = 0 # 試行回数K
        self.K_MAX = 1000 # 試行回数の最大          (本来ならもっと増やす)
        self.A_MIN = 100 # 非対角行列の要素の最小値 (本来なら0.001ぐらいの値にする)
    
    def getTheta(self,row,column):
        '''P_kに与える行列を得る'''
        tmp = self.matrix[column][column] - self.matrix[row][row]
        if tmp == 0 : return np.pi/4
        result = np.arctan(2*self.matrix[row][column]/tmp)
        result/= 2
        return result
    def getAbsMax(self):
        '''
        行列（ndarry)のなかの非対角成分で絶対値が最大のインデックス(配列番号)を返す  
        @return cdn (list) 絶対値最大のij >> [i,j]
        '''
        tmp = np.abs(self.matrix)
        l = self.row
        for i in range(l):
            tmp[i][i] = 0
        arg_max = np.argmax(tmp)
        a = arg_max // l
        b = arg_max % l
        cdn = [a,b]
        return cdn
    
    def createMatrixP(self,i,j,theta):
        '''
        i,j,行列についてのP行列を算出する
        @param i,j : i行（番号）j列（番号）
        @return ndarry P行列
        '''
        listP =[[1 if i==j else 0 for i in range(196)] for j in range(196)]
        listP[i][i] = np.cos(theta)
        listP[i][j] = np.sin(theta)
        listP[j][i] = -1 * np.sin(theta)
        listP[j][j] = np.cos(theta)
        ndarrayP = np.array(listP)
        return ndarrayP

    def nextMatrix(self,i,j,theta):
        '''
        Pk^-1 A P の計算
        '''
        # tmp = [x[:] for x in self.matrix] # 2次元配列の値渡し
        tmp = np.copy(self.matrix)
        co = np.cos(theta)
        si = np.sin(theta)
        for k in range(DIM):
            self.matrix[i][k] = tmp[i][k] * co - tmp[j][k] * si
            self.matrix[j][k] = tmp[i][k] * si + tmp[j][k] * co
            self.matrix[k][i] = tmp[k][i] * co - tmp[k][j] * si
            self.matrix[k][j] = tmp[k][i] * si + tmp[k][j] * co

    
    def checkMatrixLambda(self,threshold):
        '''
        全要素が閾値以下かを調べる
        @param   閾値
        @return  閾値以下でない：false 閾値以上：true
        '''
        for i in range(self.row):
            for j in range(i+1,self.column):
                if abs(self.matrix[i][j]) >= threshold :
                    return False
        return True
    
    # ヤコビ行列から固有値・固有ベクトルを求める
    def JacobiMethod(self):
        while True:
            maxdata = self.getAbsMax()
            i = maxdata[0]
            j = maxdata[1]
            print(str(self.k)+" "+str(self.matrix[i][j]))
            if abs(self.matrix[i][j]) < self.A_MIN:
                break
            theta = self.getTheta(i,j)
            # P_k = self.createMatrixP(i,j,theta)
            # self.matrix = np.dot(P_k.T,self.matrix)
            # self.matrix = np.dot(self.matrix,P_k)
            # self.nextMatrix(i,j,theta)
            ###############################
            tmp = np.copy(self.matrix)
            co = np.cos(theta)
            si = np.sin(theta)
            for k in range(DIM):
                self.matrix[i][k] = tmp[i][k] * co - tmp[j][k] * si
                self.matrix[j][k] = tmp[i][k] * si + tmp[j][k] * co
                self.matrix[k][i] = tmp[k][i] * co - tmp[k][j] * si
                self.matrix[k][j] = tmp[k][i] * si + tmp[k][j] * co
            ###############################
            self.k+=1
            if self.k > self.K_MAX:
                break

def main():
    for ctgry in range(CTGRY):
        # 共分散行列の読み込み
        sigma = fileRead_sigma('./sigma/sigma{0:02d}.txt'.format(ctgry+1))
        # 全ての対角成分を0に近づける
        narrayA = np.array(sigma)
        matA = Jacobi(narrayA)
        matA.JacobiMethod()
        print(matA.k)
        print(matA.matrix)
        out = list(matA.matrix)
        # 出力
        wfname = './test.txt'
        fileWrite(wfname,out)

if __name__ == '__main__':
    main()

File: KE2/latter term/kadai3.java
import java.io.*;



public final int DIM = 196;
public final int DATASIZE = 180;
public final int  CTGRY = 1;
public final int K_MAX = 10000;
public final double A_MIN = 100;


private class Matrix{
    public double[][] matrix;
    public int row;
    public int column;
    
    public Matrix(double[][] input) {
        this.row = input.length;
        this.column = input[0].length;
        this.matrix = new double[this.row][this.column];
        for (int i = 0; i < input.length; i++) {
            for (int j = 0; j < input[0].length; j++) {
                this.matrix[i][j] = input[i][j];
            }
        }
    }
    /**
	 * 行数・列数を指定して初期化
	 * @param Row
	 * @param Column
	 */
	public Matrix(int Row,int Column){
		this.row       = Row;
		this.column    = Column;
		this.matrix = new double[this.row][this.column];
	}
    /**
	 * 行列の積算が可能か判定する(X × Y)
	 * @param x	{Matrix} 掛けられる行列
	 * @param y	{Matrix} 掛ける行列
	 * @return `{boolean}
	 */
	public static boolean multipliable(Matrix x ,Matrix y){
		if(x.column!=y.row)return false;
		return true;
    }
    
    /**
	 * 行列同士・行列とベクトルの積を計算する（[matrix] this * [matrix] target
	 * @param target {Matrix} 掛けられる行列orベクトル
	 * @return result {Matrix} 積（行列）
	 */
	public Matrix multiplyMatrix(Matrix target){
		// 掛けられる行列の列数と掛ける行列の行数が等しいなら
		Matrix result = new Matrix();
		if(multipliable(this,target)){
			// 積の計算処理を実装せよ
			result.matrix = new double [this.row][target.column];
			for(int k=0;k<this.row;k++){
				for(int n=0;n<target.column;n++){
					double resultElement = 0.0;
					for(int i=0;i<target.row;i++){
						resultElement+=this.matrix[k][i] * target.matrix[i][n];
					}
					result.matrix[k][n] = resultElement;
				}
			}
			result.row = result.matrix.length;
			result.column = result.matrix[0].length;
        }else{
            System.out.println("[Matrix.multiplyMatrix] 要素数が計算できる組み合わせとなっていません");
            System.exit(0);
        }
		// 積の結果をMatrix型で返す
		return result;
    }
    /**
	 * 行列同士・行列とベクトルの積を計算する（[matrix] a * [matrix] b
	 * @param a {Matrix} 掛ける行列orベクトル
	 * @param b {Matrix} 掛けられる行列orベクトル     * 
	 * @return result {Matrix} 積（行列）
	 */
	public Matrix multiplyMatrix2(Matrix a,Matrix b){
		// 掛けられる行列の列数と掛ける行列の行数が等しいなら
		Matrix result = new Matrix();
		if(multipliable(a,b)){
			// 積の計算処理を実装せよ
			result.matrix = new double [a.row][b.column];
			for(int k=0;k<a.row;k++){
				for(int n=0;n<b.column;n++){
					double resultElement = 0.0;
					for(int i=0;i<b.row;i++){
						resultElement+=a.matrix[k][i] * b.m[i][n];
					}
					result.matrix[k][n] = resultElement;
				}
			}
			result.row = result.matrix.length;
			result.column = result.matrix[0].length;
        }else{
            System.out.println("[Matrix.multiplyMatrix] 要素数が計算できる組み合わせとなっていません");
            System.exit(0);
        }
		// 積の結果をMatrix型で返す
		return result;
	}
    /**
	 * thisの転置行列を返すメソッド
	 * @return result {Matrix} 転置行列
	 */
	public Matrix transpose(){
		Matrix result = new Matrix(this.column,this.row);
		for(int i=0;i<this.row;i++){
			for(int j=0;j<this.column;j++){
				result.matrix[j][i] = this.matrix[i][j];
			}
		}
		return result;
	}
}

private class Jacobi extends Matrix{
    protected int k;
    protected int K_MAX;
    protected int A_MIN;

    public Jacobi(double[][] input) {
        super(input);
        this.k = 0;
        this.K_MAX = 10000;
        this.A_MIN = 1;
    } 

    public int[] getAbsMax(){
        double tmp = 0;
        int cdn[] = {0,0};
        for(int i=0;i<this.row;i++){
            for(int j=i+1;j<this.columns;j++){
                if(Math.abs(this.matrix[i][j]) > tmp){
                    tmp = Math.abs(this.matrix[i][j]);
                    cdn[0] = i;
                    cdn[1] = j;
                }
            }
        }
        return cdn;
    }

    public double getTheta(int i,int j){
        double result = Math.atan(2*this.matrix[i][j]/(this.matrix[j][j]-this.matrix[i][i]));
        result /= 2;
        return result;
    }

    public Matrix createMatrixP(int row,int col){
        double[][] Pk = new double[DIM][DIM];
        for(int i=0;i<DIM;i++){
            for(int j=0;j<DIM;j++){
                if(i==j)Pk[i][j] = 1;
                else Pk[i][j] = 0;
            }
        }
        double theta = this.getTheta(row,col);
        Pk[row][row] = Math.cos(theta);
        Pk[row][col] = Math.sin(theta);
        Pk[col][row] = -1 * Math.sin(theta);
        Pk[col][col] = Math.cos(theta);
        Matrix matPk = new Matrix(Pk);
        return matPk;
    }

    public boolean checkMatrixLambda(double threshold){
        for(int i=0;i<this.row;i++){
            for(int j=i+1;j<this.column;j++){
                if(Math.abs(this.matrix[i][j]) >= threshold){
                    return false;
                }
            }
        }
        return true;
    }

    public void JacobiMethod(){
        while(true){
            double[] maxdata = this.getAbsMax();
            int i = parseInt(maxdata[0]);
            int j = parseInt(maxdata[1]);
            if(Math.abs(this.matrix[i][j])<this.A_MIN){
                break;
            }
            Matrix p_k = this.createMatrixP(i, j);
            this = this.multiplyMatrix(p_k.transpose(),this);
            this = this.multiplyMatrix(this,p_k);
            this.k+=1;
            if(this.k > this.K_MAX){
                break;
            }
        }
    }
}

public class kadai3{
    

    public double[][] fileRead_sigma(String fname){
        try{
            BufferedReader bufferedReader = new BufferedReader(new FileReader(fname));
            double[][] sigma = new double[DIM][DIM];
            for(int i=0;i<this.DIM;i++){
                String[] line = bufferedReader.readLine().split("\t",0);
                for(int j=0;j<this.DIM;j++){
                    sigma[i][j] = parseDouble(line[j]);
                }
            }
            fileReader.close();
            return sigma;
        }catch(IOException e){
            e.printStackTrace();
            System.exit(0);
        }
    }

    public static void main(String[] args) {
        for(int i=0;i<CTGRY;i++){
            // * ファイル名選択
            String fname;
            if(i<10)  fname= "./sigma/sigma"+0+i;
            else fname = "./sigma/shigma"+i;
            double[][] sigma =this.fileRead_sigma(fname);
            Jacobi matA = new Jacobi(sigma);
            matA.JacobiMethod();
        }
	}
}
